### PR TITLE
Add support for VIRTUAL_FILE

### PR DESCRIPTION
**Pull request by ben-clayton**
`VIRTUAL_FILE` declares a file in a new virtual file system.
These files can now be referenced by the `SHADER` commands instead of using inline source.

The DXC compiler will first look in the virtual file system for any `#include`s before falling back to the standard file system.

This allows us to write tests that exercise the compiler and debugger handling of multiple files.

Added `relative_includes_hlsl.amber` which checks that DXC correctly includes relative to the current file, not the root file.

### CODE REVIEW

File: docs/amber_script.md [hunk 1]
```diff
@@ -68,8 +68,63 @@ set of data types.
...
+
+```groovy
+VIRTUAL_FILE {path}
+ {file-content}
+END
+```
+
+Paths must be unique.
+
+Shaders can directly reference these virtual files for their source. \
```
Comment (dj2) [inline docs/amber_script.md:87]: What's with the \'s at the end of some of these?

File: src/amberscript/parser.cc [hunk 2]
```diff
@@ -370,19 +372,42 @@ Result Parser::ParseShaderBlock() {
...
+      return r;
 
-  shader->SetData(data);
+    auto path = token->AsString();
 
-  token = tokenizer_->NextToken();
-  if (!token->IsIdentifier() || token->AsString() != "END")
-    return Result("SHADER missing END command");
+    std::string data;
+    r = script_->GetVirtualFiles()->Get(path, &data);
```
Comment (dj2) [inline src/amberscript/parser.cc:390]: What about just `script_->GetVirtualFile(path, &data);` keep the implementation hidden

File: src/amberscript/parser.cc [hunk 3]
```diff
@@ -2896,5 +2921,24 @@ Result Parser::ParseTolerances(std::vector<Probe::Tolerance>* tolerances) {
   return {};
 }
 
+Result Parser::ParseVirtualFile() {
+  auto token = tokenizer_->NextToken();
+  if (!token->IsIdentifier() && !token->IsString())
+    return Result("invalid virtual file path");
+  auto path = token->AsString();
```
Comment (dj2) [inline src/amberscript/parser.cc:2928]: nit: add blank line above

File: src/amberscript/parser.cc [hunk 4]
```diff
@@ -2896,5 +2921,24 @@ Result Parser::ParseTolerances(std::vector<Probe::Tolerance>* tolerances) {
...
+  if (!r.IsSuccess())
+    return r;
+
+  auto data = tokenizer_->ExtractToNext("END");
+
+  token = tokenizer_->NextToken();
+  if (!token->IsIdentifier() || token->AsString() != "END")
+    return Result("VIRTUAL_FILE missing END command");
+
+  return script_->GetVirtualFiles()->Add(path, data);
```
Comment (dj2) [inline src/amberscript/parser.cc:2940]: `script_->SetVirtualFile(path, data)`

File: src/amberscript/parser_shader_test.cc [hunk 5]
```diff
@@ -235,6 +235,26 @@ END)";
   EXPECT_EQ("2: extra parameters after SHADER command: INVALID", r.Error());
 }
```
Comment (dj2) [inline src/amberscript/parser_shader_test.cc:237]: Add a test for duplicate virtual file name

File: src/script.cc [hunk 6]
```diff
@@ -18,7 +18,7 @@
 
 namespace amber {
 
-Script::Script() = default;
+Script::Script() : virtual_files_(new VirtualFileStore()) {}
```
Comment (dj2) [inline src/script.cc:21]: nit: `virtual_files_(std::make_unique<VirtualFileStore>())`

File: src/virtual_file_store.cc [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "src/virtual_file_store.h"
+
+namespace amber {
+namespace {
+
+bool HasPrefix(const std::string& str, const std::string& prefix) {
+  return str.rfind(prefix, 0) == 0;
```
Comment (dj2) [inline src/virtual_file_store.cc:21]: Why `rfind`? Is this just `str.compare(0, prefix.size(), prefix);`

File: src/virtual_file_store.h [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+      return Result("Virtual file '" + path + "' already declared");
+    }
+    files_by_path_.emplace(canonical, content);
+    return {};
+  }
+
+  /// Look up the virtual file by path. If the file was found, the content is
+  /// assigned to content.
+  Result Get(const std::string& path, std::string* content) const {
+    auto canonical = GetCanonical(path);
```
Comment (dj2) [inline src/virtual_file_store.h:48]: Should we either `assert(content)` or check and return failure if it's nullptr?

File: docs/amber_script.md (hunk 1 above)
Comment (ben-clayton) [reply docs/amber_script.md:87]: Forces a new line in markup. Double newline starts a new paragraph (larger blank space)

File: src/amberscript/parser.cc (hunk 2 above)
Comment (ben-clayton) [reply src/amberscript/parser.cc:390]: `VirtualFileStore` is handed down to the compilers in places the full script is not available. Also gives me a clean place to add the virtual file tests, without cluttering the script further.

File: src/amberscript/parser.cc (hunk 3 above)
Comment (ben-clayton) [reply src/amberscript/parser.cc:2928]: Ack.

File: src/amberscript/parser.cc (hunk 4 above)
Comment (ben-clayton) [reply src/amberscript/parser.cc:2940]: See above. I could add helpers to the script to do forward on to the `VirtualFileStore` if you prefer?

File: src/amberscript/parser_shader_test.cc (hunk 5 above)
Comment (ben-clayton) [reply src/amberscript/parser_shader_test.cc:237]: Ack

File: src/amberscript/parser.cc (hunk 2 above)
Comment (dj2) [reply src/amberscript/parser.cc:390]: That's fine, and `Get` and `Add` make sense as the API into the virtual file system. I'm just thinking if the parser should care, or if it should just tell the script has files to add/get.

File: src/script.cc (hunk 6 above)
Comment (ben-clayton) [reply src/script.cc:21]: `make_unique` was introduced in C++14. I thought amber targeted 11?

File: src/amberscript/parser.cc (hunk 4 above)
Comment (dj2) [reply src/amberscript/parser.cc:2940]: Yea, that's what I was thinking.

File: src/script.cc (hunk 6 above)
Comment (dj2) [reply src/script.cc:21]: Been working in tint too long, `virtual_files_(MakeUnique<VirtualFileStore>())`

Comment (ben-clayton) [reply src/script.cc:21]: Ack.

File: src/virtual_file_store.h (hunk 8 above)
Comment (ben-clayton) [reply src/virtual_file_store.h:48]: Do you mean zero length? I could certainly do this in `Set`. I think the `Get` handler in this case is probably fine (it won't be found).

File: src/virtual_file_store.cc (hunk 7 above)
Comment (ben-clayton) [reply src/virtual_file_store.cc:21]: No good reason. First example I found on the web. Obviously not a good one.

File: src/virtual_file_store.h (hunk 8 above)
Comment (dj2) [reply src/virtual_file_store.h:48]: Content is a pointer, so what happens if I call `Get("foo", nullptr)` ?

File: src/amberscript/parser.cc (hunk 2 above)
Comment (ben-clayton) [reply src/amberscript/parser.cc]: Done.

File: src/amberscript/parser.cc (hunk 3 above)
Comment (ben-clayton) [reply src/amberscript/parser.cc:2929]: Done.

File: src/amberscript/parser.cc (hunk 4 above)
Comment (ben-clayton) [reply src/amberscript/parser.cc]: Done.

File: src/amberscript/parser_shader_test.cc (hunk 5 above)
Comment (ben-clayton) [reply src/amberscript/parser_shader_test.cc:237]: Done. Also added one for blank paths.

File: src/script.cc (hunk 6 above)
Comment (ben-clayton) [reply src/script.cc]: Done.

File: src/virtual_file_store.cc (hunk 7 above)
Comment (ben-clayton) [reply src/virtual_file_store.cc]: Done.

File: src/virtual_file_store.h (hunk 8 above)
Comment (ben-clayton) [reply src/virtual_file_store.h:56]: Done on both.

Comment (ben-clayton) [reply src/virtual_file_store.h:56]: Oh you mean `content`. Yeah, could `assert()` - not sure it'll be _that much_ clearer than the null-pointer segfault you'll get. :)

If this is a standard thing done in amber for out variables, I can add.

Comment (ben-clayton) [reply src/virtual_file_store.h:58]: I saw some other places that does indeed `assert()` on out parameters. Done.

Comment (dj2) [reply src/virtual_file_store.h:58]: I don't remember what the standard is in amber, but `assert(content && "content should not be nullptr")` can be clearer, heh

**Comment by ben-clayton** [comment]
Thanks. Will address style issue and anything else that's flagged by presubmits.

File: src/virtual_file_store.h (hunk 8 above)
Comment (ben-clayton) [reply src/virtual_file_store.h:58]: I've seen some compilers warn about `&&`'ing with string literals, so I'll opt to keep as a basic `assert(content)` if that's okay.

Comment (dj2) [reply src/virtual_file_store.h:58]: Works for me.

**Comment by ben-clayton** [comment]
`CI-macos-clang-debug` failed on:
* `./tests/cases/copy_format_buffer.amber`
* `./tests/cases/magfilter_nearest.amber`

With the test failures:

```
[ VUID-vkFreeMemory-memory-00677 ] Object: 0x7f874efb3808 (Type = 3) | Cannot call vkFreeMemory on VkDeviceMemory 0x1a000000001a[] that is currently in use by a command buffer. The Vulkan spec states: All submitted commands that refer to memory (via images or buffers) must have completed execution (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-vkFreeMemory-memory-00677)
[ERROR] validation layer (Validation):
[ VUID-vkDestroyBuffer-buffer-00922 ] Object: 0x190000000019 (Type = 9) | Cannot free VkBuffer 0x190000000019[] that is in use by a command buffer. The Vulkan spec states: All submitted commands that refer to buffer, either directly or via a VkBufferView, must have completed execution (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-vkDestroyBuffer-buffer-00922)
[ERROR] validation layer (Validation):
[ VUID-vkDestroyRenderPass-renderPass-00873 ] Object: 0x7f874efb3808 (Type = 3) | Cannot call vkDestroyRenderPass on VkRenderPass 0x60000000006[] that is currently in use by a command buffer. The Vulkan spec states: All submitted commands that refer to renderPass must have completed execution (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-vkDestroyRenderPass-renderPass-00873)
[ERROR] validation layer (Validation):
...
```

@dj2 thinks that this is a timeout teardown issue, which makes sense, as this particular CI bot is running **really** slowly.

I see nothing to suggest that this change has affected the build state, so @dj2 suggests landing this PR, and following up with suppressing/fixing the flaky tests.
